### ISSUE TITLE
Tracking issue for `proc_macro::Span` inspection APIs

### ISSUE THREAD

**Issue by alexcrichton** [rust-lang/rust#54725]
This issue is intended to track a number of unstable APIs which are used to inspect the contents of a `Span` for information like the file name, byte position, manufacturing new spans, combining them, etc.

This issue tracks the `proc_macro_span` unstable feature.

I'm not currently sure what the blockers are for stabilization, but I know there are some! If others would like to fill this in that'd be great!

---

Public API:

```rust
impl Span {
    pub fn source_file(&self) -> SourceFile;
    pub fn start(&self) -> LineColumn;
    pub fn end(&self) -> LineColumn;

    pub fn source_text(&self) -> Option<String>;

    pub fn eq(&self, other: &Span) -> bool;

    pub fn join(&self, other: Span) -> Option<Span>;

    pub fn parent(&self) -> Option<Span>;
    pub fn source(&self) -> Span;
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub struct LineColumn {
    pub line: usize,
    pub column: usize,
}

impl !Send for LineColumn {}
impl !Sync for LineColumn {}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct SourceFile(..);

impl SourceFile {
    pub fn path(&self) -> PathBuf;
    pub fn is_real(&self) -> bool;
}

impl Literal {
    pub fn subspan<R: RangeBounds<usize>>(&self, range: R) -> Option<Span>;
}

impl Group {
    pub fn span_open(&self) -> Span;
    pub fn span_close(&self) -> Span;
}
```

**Comment by softprops** [comment]
I wanted to shed some pain this is causing with [tarpaulin](https://crates.io/crates/cargo-tarpaulin).  

Tarpaulin has worked amazingly well for me and my company as a replacement for kcov which historically been a pain to get accurate, reliable and correct results. At the moment tarpaulin stands as the most promising goto option for codecoverage in rust and just feels more like the only first class tooling option. 

Having one of those when choosing to invest in a technology is important for many company's adoption story for checking off code quantity checkmarks. When they see that rust doesn't have a reasonable code quality story that works on  stable rust, that's can result in a "pass" rather than "I'll take it". There are currently some work arounds for making this work-ish on stable but it feels very much like the story serde was in a year or so ago when you wanted to show all your friends how amazing serde was but then was embarrassed to show what it took to make work on stable because of a macro stabilization blocker.

**Comment by JakeTherston** [comment]
With procedural macros having reached a point where they're very useful on stable, I expect many users will find themselves needing access to this information. Would it be reasonable to only stabilize parts of the `Span` API that are not too risky? Perhaps exposing a function that optionally returns the path of the file where a macro is invoked if such a file exists?

**Comment by matklad** [comment]
@davidlattimore this is fascinating, but slightly off-topic for the issue. I've created a thread on internals: https://internals.rust-lang.org/t/macros-vs-incremental-parsing/9323

**Comment by est31** [comment]
The [column!()](https://doc.rust-lang.org/stable/std/macro.column.html) macro as well as [std::panic::Location::column](https://doc.rust-lang.org/stable/std/panic/struct.Location.html#method.column) are returning 1-based columns while the span available from the proc-macro crate [is 0-based](https://doc.rust-lang.org/nightly/proc_macro/struct.LineColumn.html#structfield.column) according to its docs. Is this inconsistency intended?

**Comment by est31** [comment]
This thread has more discussion about 1-based columns: https://github.com/rust-lang/rust/pull/46762#issuecomment-352474639

**Comment by est31** [comment]
Another open question is how this API relates to https://github.com/rust-lang/rust/issues/47389 which is about minimizing span information throughout the compiler. Should stabilization be blocked until a design for #47389 is found? Is it too late already as we have `set_span` functionality? @michaelwoerister what do you think?

**Comment by michaelwoerister** [comment]
#47389 is mostly concerned about data types that are used later in the compilation pipeline, such as type information and MIR. Exposing things at the proc-macro level should not be too bad.

**Comment by est31** [comment]
But rust-analyzer might one day expand the scope of incremental compilation to the parsing stage, right?

**Comment by matklad** [comment]
That is already true for rust-analyzer, macro expansion stage runs incrementally.  We do’t yet support line/column macro, but I think what we would do is that we’ll just expand them to a dummy value like 42. This probably won’t break macro (It could be this line indeed), but will allow us to avoid re-expanding macro after every keystroke

**Comment by est31** [comment]
What we *could* do is store for each macro invocation whether it has invoked any of the inspection APIs or not and then re-run the macro only if any of the span information it has requested has had any changes. Basically only adding the infos the requested spans to the hash (or only adding span info to the hash if there's been any request of span info). However, as this has to update state, it would make span inspection slower than it could be, especially if something inspects spans routinely. So not sure about the performance impacts of this.

**Comment by jhpratt** [comment]
@alexcrichton @matklad Any update on what blockers are for this issue? It doesn't seem like there should be anything holding this back.

**Comment by est31** [comment]
@jhpratt my question about 1-based vs 0-based is still not answered: https://github.com/rust-lang/rust/issues/54725#issuecomment-497246753

**Comment by jhpratt** [comment]
Hm, that's a fairly simple one. My intuition would certainly lean towards 1-based.

**Comment by syntacticsugarglider** [comment]
Is this still blocked on questions with regard to the inconsistency between two base indices?

**Comment by Alxandr** [comment]
I just thought I'd chime in here, cause I just ran into an issue where I would like to be able to map "place in code" to something else. Basically, I don't care about files or columns or indexes or anything, I just need something that's `Eq + Hash`, and that I can get 2 from (start and end) for a `Span` (also that it follows that any given token t has the same end as the start of the next token). This could just be a usize that's local to the current macro invocation and doesn't have to mean anything to the broader world, but it gives the ability to sort and map on "code points".

**Comment by de-vri-es** [comment]
I'd love to see this stabilized too. In particular, [`assert2`](https://docs.rs/assert2) uses `Span::source_text()` to show the expression that caused an assertion failure:

![](https://raw.githubusercontent.com/de-vri-es/assert2-rs/2db44c46e4580ec87d2881a698815e1ec5fcdf3f/pattern-match.png)

 On stable it falls back to regular stringification, but that certainly looks less pretty.

**Comment by Aaron1011** [comment]
Note that `Span::parent` allows you to write 'extra unhygenic' macros, which can see identifiers in a way that would otherwise be impossible:

```rust
// src/lib.rs
#![feature(proc_macro_span)]

use proc_macro::TokenStream;

#[proc_macro]
pub fn no_hygiene(tokens: TokenStream) -> TokenStream {
    tokens.into_iter().map(|mut tree| {
        let mut span = tree.span();
        while let Some(parent) = span.parent() {
            span = parent;
        }
        tree.set_span(tree.span().resolved_at(span));
        tree
    }).collect()
}
```

```rust
use no_hygiene::no_hygiene;

macro_rules! weird {
    () => {
        println!("{}", no_hygiene!(my_ident));
    }
}

fn main() {
    let my_ident = "Hello";
    weird!();
}
```

Here, `weird!()` is able to access `my_ident` via `no_hygiene!`. This is already possible with just `Span::resolved_at` in some circumstances - if a single token has a `Span` with the desired hygiene, a proc-macro can construct arbitrary tokens with the desired hygiene. However, `Span::parent` allows hygiene to be completely bypassed - a proc-macro can see into a particular scope that it would otherwise be completely isolated from.

Using `Span::parent` may be useful for the purposes of emitting error messages, but I'm worried about macros generating working code that intentionally uses this. This seems like an easy way to end up with code that has confusing non-local effects.

**Comment by Aaron1011** [comment]
We might want to gate 'read-only' APIs (e.g. `Span::source_text`, `Span::start`) under a separate feature, since they could be stabilized without increasing the power of proc macros w.r.t hygiene.

**Comment by udoprog** [comment]
So I'm looking into what's needed to get the following stabilized in `libproc_macro`:

* `Span::start`
* `Span::end`
* `LineColumn`

There are two outlined concerns that I've been able to digest so far. If someone has more, I'd love to hear them.

* [This question by @est31](https://github.com/rust-lang/rust/issues/54725#issuecomment-497246753) whether the column should be switched to be 1-based, instead of 0-based to be consistent with how other APIs report columns.
* [This concern raised by @matklad](https://github.com/rust-lang/rust/issues/54725#issuecomment-459722888) about exposing absolute span information in general, since they could have an effect on the output of a macro.

#### 1-based columns

The corresponding functions and `LineColumn` structure in proc_macro2 are already marked as stable with the same 0-column specification. But as pointed out in alexcrichton/proc-macro2#237, while they are available they only supply dummy values on stable right now. But if desired, I'd be happy to put in the work to modify the reported column information to be 1-based instead. I personally don't have a strong opinion either way. The manner in which I've used spans in [genco](https://github.com/udoprog/genco) (and why I now have a desire for them to be stabilized :)) only cares about relative spans.

#### reported spans affecting the output of proc-macros

Given that proc-macros today can execute arbitrary code, and [this seems like something we just have to live with](https://github.com/launchbadge/sqlx#compile-time-verification). I don't know how beneficial limiting the existing proc_macro API surface with the goal of making macro execution as reproducible as possible is at this point. I personally wouldn't want to block on this. But for my specific use-case, changing the spans to be reported in a manner which is local to the macros themselves would also give me what I need. So if desired, I'd be for implementing this change.

To clarify, that means this (ish):
```
/// A comment
macro! { hello // start {column: 8, line: 2}
    world // start {column: 4, line: 3}
};
```
Would instead be reported in the macro as this:
```
/// A comment
macro! { hello // start {column: 1, line: 1}
    world // start {column: 4, line: 2}
};
```

**Comment by matklad** [comment]
>reported spans affecting the output of proc-macros
>I don't know how beneficial limiting the existing proc_macro API surface with the goal of making macro execution as reproducible as possible is at this point.

I don't think there are any concerns about *reproducibility*, there are concerns about *incremental compilation* which is something completely different. I think its important that we get opinion of some incremental compilation expert from t-compiler about this (cc @wesleywiser? not really sure whom to CC). 

Are we ok with exposing absolute position information to macros? AFAIUI, that means that either:

* we'll have to re-expand *all* proc-macros even for trivial changes like addition of blank line.
* we'll have to record "has this macro looked at line/column?" bit of info, and *still* re-expand those specific macros for trivial changes.

**Comment by udoprog** [comment]
> I don't think there are any concerns about _reproducibility_, there are concerns about _incremental compilation_ which is something completely different.

So when I phrase it as the output of a macro being reproducible, that is what I'm referring to. Given a collection of inputs it would produce an identical output which would allow it to be cached during incremental compilation.

But to my wider point, a proc-macro can generate random numbers, access the filesystem, or connect to a database (the example I linked to). I'd be very curious to hear about plans for incremental compilation, but given these facts I have a hard time envisioning an approach that allows us to make decisions about re-expanding beyond some form of documented best-effort heuristic.

**Comment by est31** [comment]
> But to my wider point, a proc-macro can generate random numbers, access the filesystem, or connect to a database (the example I linked to).

That's possible for now, but it's not a documented feature that proc macros are unsandboxed and it's likely that wasm based macros won't get that capability.

To the general question of incremental compilation see also https://github.com/rust-lang/rust/issues/54725#issuecomment-502932263

**Comment by wesleywiser** [comment]
@matklad While I do know a bit about the incremental system, I wouldn't consider myself an expert. I believe @pnkfelix knows a fair bit more than I.

**Comment by CreepySkeleton** [comment]
I'm specifically interested in `Span::join` that would allow for painless proc-macro error messages on stable. There's [the hack](https://docs.rs/syn/1.0.23/src/syn/error.rs.html#219-238) that makes it already possible, but in order to exploit it, one has to carry around two spans: the span to the first token and the span to the last token.

If `join` were available, `syn` could do the joining in its `ast.span()` methods and the two-spans trick would become easier to pull off since the joined span had already covered the whole range.

**Comment by Aaron1011** [comment]
> I'm specifically interested in Span::join that would allow for painless proc-macro error messages on stable

Before `Span::join` is stabilized, we should decide how it should handle hygiene. If I write `first.join(second)` should the returned `Span` use `Span::resolved_at(first)` or `Span::resolved_at(second)`? Or should we change the return type to `Result<Span, JoinErr>`, and have `enum JoinError { DifferentFile, DifferentHygiene  }`?

**Comment by SergioBenitez** [comment]
> Note that `Span::parent` allows you to write 'completely unhygenic' macros, which can see identifiers in a way that would otherwise be impossible:

Not to advocate in either direction, but this is already possible today, without `Span::parent()`:

```rust
// main.rs
mod submod {
    macro_rules! make_local_weird {
        ($local:item) => (
            #[macro_export]
            macro_rules! weird {
                () => (
                    println!("{}", export! { $local private })
                )
            }

            pub use weird;
        )
    }

    make_local_weird!(fn f() {});
}

pub fn main() {
    let private = "hi";
    submod::weird!();
}
```

```rust 
// lib.rs
use proc_macro::TokenStream;

use syn::spanned::Spanned;
use syn::{parse_macro_input, Result};
use syn::parse::{Parse, ParseStream};
use quote::quote;

struct Export {
    item: syn::Item,
    ident: syn::Ident,
}

impl Parse for Export {
    fn parse(input: ParseStream) -> Result<Self> {
        Ok(Export {
            item: input.parse()?,
            ident: input.parse()?,
        })
    }
}

#[proc_macro]
pub fn export(input: TokenStream) -> TokenStream {
    let input = parse_macro_input!(input as Export);
    let mut ident = input.ident;
    let span = ident.span().resolved_at(input.item.span());
    ident.set_span(span.into());
    quote!(#ident).into()
}
```

**Comment by jonas-schievink** [comment]
> Are we ok with exposing absolute position information to macros? AFAIUI, that means that either:
> 
>    * we'll have to re-expand _all_ proc-macros even for trivial changes like addition of blank line.
> 
>    * we'll have to record "has this macro looked at line/column?" bit of info, and _still_ re-expand those specific macros for trivial changes.

A third option would be to provide dummy information to proc macros from rust-analyzer. AFAIK, most proc macros use the data they obtain from these APIs only at runtime, and do not really make any decisions based on it at expansion time (and even then, they should work fine if line/column is always 1, unless they're being actively annoying).

**Comment by udoprog** [comment]
> A third option would be to provide dummy information to proc macros from rust-analyzer. AFAIK, most proc macros use the data they obtain from these APIs only at runtime, and do not really make any decisions based on it at expansion time (and even then, they should work fine if line/column is always 1, unless they're being actively annoying).

The macros requiring this information appear to do so for whitespace information ([mine does](https://github.com/udoprog/genco), and so does done of the linked issues above as well). I don't see how this would work without a difference between spans. So being actively annoying might just be a valid use case.

**Comment by aticu** [comment]
Just to give another example of proc-macros requiring white-space information to function properly: [inline-python](https://crates.io/crates/inline-python)

Also another use-case I could think of would be generating documentation based on Rust code in the macro input.
Without whitespace information, this could look quite bad:

```rust
std :: mem :: ManuallyDrop < T > :: drop ( & t ) ;
// vs
std::mem::ManuallyDrop<T>::drop(&t);
```

```rust
call ( & x [ 0 ] ) ;
// vs
call(&x[0]);
```

Though the second use-case could probably handle invalid line/column data fairly well, if it is only given when executed by an IDE. In the worst case it would possibly just generate unreadable documentation that isn't used by the IDE.

**Comment by jan-hudec** [comment]
Since these examples only need relative position of the tokens, the fourth option is to provide the information, but count lines from the start of the macro call only (so the first token in the token stream always has line = 1). That way if whitespace changes inside the macro call it needs to be re-expanded, but if anything changes before it it does not.

**Comment by m-ou-se** [comment]
> A third option would be to provide dummy information to proc macros from rust-analyzer. AFAIK, most proc macros use the data they obtain from these APIs only at runtime, and do not really make any decisions based on it at expansion time

[`ct_python!{}`](https://docs.rs/ct-python/) executes Python code at compile time. Changing the span information could change the meaning of the Python code. (Though one could definitely argue that this whole crate is a horrible idea in combination with RLS/rust-analyzer (or just in general ^^'). With RLS, it executes the Python code *while you are typing it*. :grimacing:)

Maybe it'd be useful if a proc macro can tell it is being executed by a language server instead of the compiler? Not sure what it could do better in that case though, because the language server still needs the expansion to do the rest of its job.

**Comment by jonas-schievink** [comment]
We had to rely on `Span`'s `Debug` impl because the line/column access methods are still unstable: https://github.com/knurling-rs/defmt/issues/344

**Comment by udoprog** [comment]
So tentatively it seems like the proposed blockers for stabilizing a subset of `proc_macro_span` as I detailed in https://github.com/rust-lang/rust/issues/54725#issuecomment-647140141 (i.e. `min_proc_macro_span`) have either been discussed or no one has raised much interest in them (e.g. if spans should be reported macro-relative or be 1 or 0-based).

The biggest discussion seems to be about incremental compilation, so I'd like to take as broad a view on the topic as I can here. I hope I'm not missing something important, but if I do, please do tell!

Incremental compilation of macros as they look today based on assumptions of reproducibility (https://github.com/rust-lang/rust/issues/54725#issuecomment-502932263, https://github.com/rust-lang/rust/issues/54725#issuecomment-647142018) do not seem like they're going to be affected either positively or negatively by stabilization of proc-macro APIs which allows for different outputs depending on where the macro is located in source. As mentioned a few times above (https://github.com/rust-lang/rust/issues/54725#issuecomment-647146117, https://github.com/rust-lang/rust/issues/54725#issuecomment-736371755) the stable parts of the proc-macro system is already "contaminated" with other sources of data that can cause different outputs. This is even the case [for the currently stable API of libproc_macro](https://github.com/knurling-rs/defmt/issues/344) since the output of `fmt::Debug` prints the absolute byte span of a span in the file.

So it seems like the cat is already out of the bag here. And this might very well be addressed in future proc-macro APIs. If such an API is designed in the future it could also come with a different version of `libproc_macro`.  But "leaky macros" do regardless seem to have legitimate uses, so it seems unlikely that they would ever be fully deprecated.

So I would like to push towards moving forward with stabilization with a minimal version of `proc_macro_span` which features read-only APIs. In particular parts which have no effect on hygiene (as mentioned here https://github.com/rust-lang/rust/issues/54725#issuecomment-640169091).

**Comment by afetisov** [comment]
Having absolute line&column positions will also break autoformatting. While sensible formatting of macros is mostly impossible, since they don't need to follow the Rust syntax, at the very least we can currently assume that whitespace has no semantic meaning. This means that we can add whitespace around operators, format inner blocks if we can prove that they are valid Rust code, or make macro calls properly indented. If a macro knows its absolute position, then simply adding or removing indentation is enough to arbitrarily change its behaviour.

Granted, it's a very minor point compared to a loss of hygiene or incremental compilation, but it's still an issue. 

Any kind of tooling which assumes that whitespace has no meaning will be screwed by this feature. Note that the promise of no meaningful whitespace is currently contained in the reference manual. [Quote:](https://doc.rust-lang.org/reference/whitespace.html)

> A Rust program has identical meaning if each whitespace element is replaced with any other legal whitespace element, such as a single space character.

And according to the [reference on macros](https://doc.rust-lang.org/reference/macros.html), the macros receive only a delimited token tree, so the whitespace is also irrelevant within macros.

So stabilizing this API will be strictly a backwards incompatible change w.r.t. current Rust guarantees. While, strictly speaking, the usual guarantees are with respect to code compiling and having the same behaviour, I believe that breaking compatibility with tools isn't a thing which should be taken lightly.

**Comment by droundy** [comment]
I'll like to point out that procmacros can already parse the source code to find where they might have been invoked, and can do any number of bad things that will cause changes to users code to break under various circumstances.  Making it more convenient for them to break tooling does not itself break said tooling. Any user of a proc macro is already relying on the author of that macro to make it do reasonable things.

Getting reasonable debug information in cases where macros fail is a huge win, and I'd hate got that to be lost due to some concern that macro writers will turn evil if we give them more power.

**Comment by afetisov** [comment]
You're stating it as if it's a hypothetical problem, while there are already crates which do exactly the position-dependent processing that I'm talking about. Do remember that this can cause issues for downstream users which may not even know about such features, much less use them directly. I shouldn't know which proc macros are called by my transitive dependencies, but if a macro that I use calls one then my IDE performance will suddenly plummet.

If it's only about debug information, then surely there are better ways to provide it than directly exposing source code layout. Position-independent relative offsets solve many of the problems: if I'm forbidden to offset outside of the containing macro call, then I probably can't violate hygiene, and one doesn't need to reparse the whole crate if some formatting changes. Another option is to make line&column information into opaque objects rather than just integers. You can print them, you can store them, maybe even do some simple calculation like comparison and calculating relative offset, but you can't just make one out of thin air or extract forbidden absolute positions.

Come to think of it, why would it ever be a good idea to allow proc macros to know absolute offset and to create arbitrary spans? Surely a macro has no legal way to know the contents of the surrounding code, so why would it be granted access to it in the first place? The only thing that a user of the macro sees is the code that they directly typed inside of it, and spans in _that exact_ code should be the only thing that a macro should work with. 

A different, always available option is to write a separate source-transforming tools, rather than make those capabilities always provided by the compiler. You can use a format which is related to, but different from the Rust sources. You can use the standard syntactic libraries for writing that tool, and since you're doing all the lexing and parsing, you're free to keep any extra information that you want. For a practical example you can look at the camlp5 preprocessor for the OCaml language. But at the same time no one will accidentally depend on your special syntactic features, and the other tools will know that your format is something that they can't process without special support. I don't think pulling the kitchen sink, the fridge and the whole swimming pool into the language is necessarily a great idea.

**Comment by jonas-schievink** [comment]
> I shouldn't know which proc macros are called by my transitive dependencies, but if a macro that I use calls one then my IDE performance will suddenly plummet.

IDE performance would only be minimally affected if these `Span` APIs are implemented properly. You can already make IDE performance arbitrarily bad by just using a slow proc macro, so I don't really think this is a reason to not provide this API at all.

> if I'm forbidden to offset outside of the containing macro call, then I probably can't violate hygiene

Hygiene is independent from positional information. We could remove the problematic `Span::parent` and stabilize the `LineColumn`-related APIs.

**Comment by wycats** [comment]
> Hygiene is independent from positional information. We could remove the problematic `Span::parent` and stabilize the `LineColumn`-related APIs.

This sounds great and would make it possible to reconstruct enough information about whitespace to implement a lot of whitespace-sensitive EDSLs. Any blockers?

**Comment by dvc94ch** [comment]
> Hygiene is independent from positional information. We could remove the problematic Span::parent and stabilize the LineColumn-related APIs.

This sounds good to me. Looks like the reason https://github.com/rust-lang/rust/pull/84331/files was rejected was introducing a new feature flag. I believe if instead those api's had just been made stable, it might have been fine.

**Comment by cdmistman** [comment]
Since the information is there, I'm wondering if it's possible to expose the byte offsets of a span within a file.

For example, given this (very simple and dumb) proc macro (using `syn` and `quote` for convenience)
```rs
use quote::quote;
use quote::ToTokens;
use syn::parse_macro_input;
use syn::spanned::Spanned;
use syn::ItemFn;
use syn::Stmt;

#[proc_macro_attribute]
pub fn instrument(
	_attrs: proc_macro::TokenStream,
	input: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
	let mut input = parse_macro_input!(input as ItemFn);
	let span = input.span();
	let span = format!("{:?}", input.span());
	input
		.block
		.stmts
		.push(syn::parse2::<Stmt>(quote! { println!("span is {}", stringify!(#span)); }).unwrap());
	input.to_token_stream().into()
}
```

on the following input:
```rs
#[proc_macro::instrument]
fn main() {}
```

the output of the resulting binary is:
```
span is "#0 bytes(27..39)"
```

However, those byte offsets aren't accessible in the API. Is it possible to expose those byte offsets via getters? Maybe some `fn byte_offsets(&self) -> Range<usize>`?

**Comment by dannymcgee** [comment]
> Hi all --
> 
> We read over this in @rust-lang/lang backlog bonanza today and had a few questions. There seems to be a broad set of APIs covered under this feature, so I think it would be useful to identify what subset we need to enable particular use cases.
> 
> Reviewing the thread, the major use case seemed to be reporting diagnostics (e.g., [this comment about assert2](https://github.com/rust-lang/rust/issues/54725#issuecomment-573432969)). If that's the case, it seems like exposing
> 
> * `source_text`
> * nicer diagnostic APIs (separate tracking issue, I think)
> * and _maybe_ line/column information relative to the start of the macro
> 
> would be sufficient to cover these cases. All of these have the property that one must actually edit the _contents_ of the macro to invalidate them -- though source-text is distinct from 'pure tokens'.
> 
> Are there use cases that would be not be handled by the APIs above? (And which ones require relative information?)
> 
> If not, maybe we should break out this subset and discuss there?

I had a use case a while back that really would have benefitted from knowing the path to the source file where the macro invocation occurred. I was trying to write something like a custom module resolver — basically, I wanted to be able to "import" a source file written in some other language (like a scripting or templating language), and the macro would facilitate transpiling the import to valid Rust code. But being unable to specify relative import paths made it prohibitively inconvenient/unergonomic for my liking.

**Comment by john01dav** [comment]
> > Hi all --
> > We read over this in @rust-lang/lang backlog bonanza today and had a few questions. There seems to be a broad set of APIs covered under this feature, so I think it would be useful to identify what subset we need to enable particular use cases.
> > Reviewing the thread, the major use case seemed to be reporting diagnostics (e.g., [this comment about assert2](https://github.com/rust-lang/rust/issues/54725#issuecomment-573432969)). If that's the case, it seems like exposing
> > 
> > * `source_text`
> > * nicer diagnostic APIs (separate tracking issue, I think)
> > * and _maybe_ line/column information relative to the start of the macro
> > 
> > would be sufficient to cover these cases. All of these have the property that one must actually edit the _contents_ of the macro to invalidate them -- though source-text is distinct from 'pure tokens'.
> > Are there use cases that would be not be handled by the APIs above? (And which ones require relative information?)
> > If not, maybe we should break out this subset and discuss there?
> 
> I had a use case a while back that really would have benefitted from knowing the path to the source file where the macro invocation occurred. I was trying to write something like a custom module resolver — basically, I wanted to be able to "import" a source file written in some other language (like a scripting or templating language), and the macro would facilitate transpiling the import to valid Rust code. But being unable to specify relative import paths made it prohibitively inconvenient/unergonomic for my liking.

Seconded. I was going to write a similar reply myself -- I had a use case that was almost exactly the same, where I wanted to allow something resembling including a module that's written in another language. This isn't the only use case for a feature like this though. Anything that includes things from the file system can benefit, such as [include_dir](https://crates.io/crates/include_dir).

**Comment by 450220020** [comment]
stable version required

**Comment by madsmtm** [comment]
Another use-case noted [above](https://github.com/rust-lang/rust/issues/54725#issuecomment-761602093) that is quite different from diagnostics is getting a unique identifier to use in `link_section`/`export_name`s.

I've [opened an internals thread](https://internals.rust-lang.org/t/get-unique-identifier-inside-macro/17071) for discussing that.

**Comment by daxpedda** [comment]
@Aaron1011 closed by mistake? Can't seem to find an alternative tracking issue, a PR stabilizing it or a PR removing the feature.

**Comment by Aaron1011** [comment]
Oops, sorry about that

**Comment by Shoeboxam** [comment]
Use case: I'm working on a rust library that provides LaTex proof files adjacent to each rust source file. I'd like to use a proc-macro to enrich my public documentation with links to the .tex sources and rendered .pdfs. Without a way to retrieve the source file from the span, I need to maintain relative paths to proof files in each invocation of the proc macro.

**Comment by djc** [comment]
So this issue seems to have made pretty limited progress in the 4 years since it was first created.

libs-team folks, how do we make progress here? Who can make a decision on 0- vs 1-based position indexes? How we might subset the API so it can be stabilized in different parts? Maybe a good idea to nominate for a libs-team meeting to get some consensus on what the way forward might be?

**Comment by est31** [comment]
FTR the column is now 1 based as of #87712 , so I'd call that question resolved.

The other problem is incremental compilation. There are some possible ways to get out of that, like to make line and column location relative to the start of the macro is one such possible solution: https://github.com/rust-lang/rust/issues/54725#issuecomment-1143966021 

It has already been split up: #101991

**Comment by djc** [comment]
Okay, so #101991 is for `Span::source_text()`. I ended up here today for `Span::join()`. 

> > I'm specifically interested in Span::join that would allow for painless proc-macro error messages on stable
> 
> Before `Span::join` is stabilized, we should decide how it should handle hygiene. If I write `first.join(second)` should the returned `Span` use `Span::resolved_at(first)` or `Span::resolved_at(second)`? Or should we change the return type to `Result<Span, JoinErr>`, and have `enum JoinError { DifferentFile, DifferentHygiene }`?

I think making `join()` fallible along these lines seems reasonable. The use case for resolving a disjoint `Span` at either one of the locations isn't clear to me, and seems like it might result in surprising results?

**Comment by est31** [comment]
@Shoeboxam I see, I have misunderstood your use case.

**Comment by Qix-** [comment]
Currently looking into an early attempt at adding `Diagnostic` support to `syn` and will need an interim PR for `proc_macro2` that implements `From<proc_macro2::Span> for Span` (ref https://github.com/dtolnay/proc-macro2/issues/181#issuecomment-1345365381), however there's currently no way to construct a `SourceFile` from a `PathBuf`, and no way to to further construct a `Span` from `(SourceFile, LineCol, LineCol)` info.

Could such constructors be added? Otherwise adding _meaningful_ `Diagnostic` support (#54140) will be next to impossible.

**Comment by jhpratt** [comment]
I would love the ability to construct a span from the file, line, and column. As part of a macro I recently improved, I am already tracking the line and column internally. But it's impossible (even with nightly) to point at where the error _actually_ is. The best I can do is point at the entire string, which is not very helpful.

**Comment by kanashimia** [comment]
![IMG_20221211_140249_893](https://user-images.githubusercontent.com/56224949/206902414-8cbfd714-ba78-4f64-bba5-27ac5ce7bbb5.jpg)

<details>
  <summary>[PRESS ME] Above proc-macro implementation code</summary>
  
```rust
#![feature(proc_macro_span)]
// #![feature(proc_macro_expand)]
// #![feature(proc_macro_diagnostic)]

fn parse_wgsl(input: syn::parse::ParseStream) -> syn::Result<Vec<u32>> {
    let literal: syn::LitStr = input.parse()?;

    let inner = literal.value();
    let outer = literal.token().to_string();

    let prefix = outer
        .split_once(inner.as_str())
        .expect("failed to split string")
        .0
        .len();

    let module = naga::front::wgsl::parse_str(&inner).map_err(|e| {
        let (range, _) = e.labels().next().unwrap();

        let literal_range = range.start + prefix..range.end.max(range.start + 1) + prefix;

        let span = literal
            .token()
            .subspan(literal_range)
            .expect("failed retrieving error subspan");

        syn::Error::new(span, e.message())
    })?;

    use naga::valid::*;
    use std::error::Error;
    let info = Validator::new(ValidationFlags::all(), Capabilities::all())
        .validate(&module)
        .map_err(|e| {
            // println!("{:#?}", e);
            // println!("{:#?}", e.spans().map(|(a, b)| 1).collect::<Vec<_>>());
            // println!("{:#?}", e.location(&inner));

            let range = e.spans().next().unwrap().0.to_range().unwrap();

            let literal_range = range.start + prefix..range.end.max(range.start + 1) + prefix;

            let span = literal
                .token()
                .subspan(literal_range)
                .expect("failed retrieving error subspan");

            let mut res = e.to_string();

            let mut e = e.source();
            while let Some(source) = e {
                res.push('\n');
                res.push_str(&source.to_string());
                e = source.source();
            }

            syn::Error::new(span, res)
        })?;

    use naga::back::spv;
    let shader = spv::write_vec(
        &module,
        &info,
        &spv::Options {
            lang_version: (1, 0),
            flags: spv::WriterFlags::empty(),
            ..Default::default()
        },
        Default::default(),
    )
    .expect("failed to generate spirv");

    Ok(shader)
}

#[proc_macro]
pub fn inline_wgsl(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
    let spirv = syn::parse_macro_input!(input with parse_wgsl);

    quote::quote!([#(#spirv),*]).into()
}
```
  
</details>

@jhpratt it possible currently to show good errors with this `proc_macro_span` feature as you can see above (look, even rust-analyzer shows errors!) using a `subspan` method on the `Literal`, you can even generate literal yourself with any `Span` using a `set_span` method.

The problem is that you can't easily construct `Span` itself from a different file, there are multiple issues that ask for a similar feature, here is the most fleshed out one: https://github.com/rust-lang/rfcs/pull/3200
This one is too related but neglects the topic of actually importing a file: https://github.com/rust-lang/rust/issues/99515

Rather than producing a `Span` RFC3200 has `include_str` function that reads a file to a `Literal`, which should be enough.

Basically you can report an error in the inline code, but can't report error in the external file currently on nightly.

You can write external file as rust string literal (wrap the code in quotes), and then using `proc_macro_expand` include it, but it is an awful hack.

**Comment by jhpratt** [comment]
Ah, I was not aware of the subspan method. I thought that was only a compiler internal. That would suffice for my use case.

**Comment by est31** [comment]
The issue about constructing spans from just the filename, line and column is that spans contain more info than just those three: they also contain hygiene information, including stuff like the edition. That's why there is `subspan`.

**Comment by jhpratt** [comment]
It's worth noting that there _is_ a stable AST (among other things) being worked on currently over at [rust-marker/marker](https://github.com/rust-marker/marker).

**Comment by Qix-** [comment]
@est31 The point of syn and proc_macro2 are to be able to parse arbitrary Rust code, including in contexts outside of a proc macro.

**Comment by est31** [comment]
@Qix- you can use [proc_macro::is_available](https://doc.rust-lang.org/stable/proc_macro/fn.is_available.html) for checking if you should use the `proc_macro` crate or your own span structure. syn is widely used indeed, but that doesn't mean that it has to adapt the `Diagnostic` API in a way that requires a function to create `Span`s from the extremely limited information of just a path, file and column numbers. If you want to use the `Diagnostic` API from `proc_macro` for code that is not an input of a proc macro, that's something yet another. The compiler is not linked in at such a point, so proc_macro's APIs will panic, as described in the docs of the `is_available` function. For those use cases, I'd recommend using one of the many libraries that let you print your own diagnostics, instead of linking into the compiler.

**Comment by andrewdavidmackenzie** [comment]
I would like my macro to know the source file where it was used, on stable. 
Is there any likelyhood of this API (span.source_file() in particular) ever stabalizing?

If not, is there, or will there ever be, a way to do that on stable?

**Comment by sam0x17** [comment]
What are the current barriers to stabilization? Four years ago it seemed like it was just around the corner. Would be happy to pitch in if there is a specific blocker.

**Comment by nic-hartley** [comment]
Really hurting for the lack of this right now. Also happy to contribute what I can to get this unblocked.

**Comment by m-ou-se** [comment]
We discussed this in the libs meetup today. Those of us present were happy to expose line and column information for spans, but were not yet convinced by the current unstable API.

Specifically, we might want to add a `byte_offset` in addition to the `line` and `column` information, which would be hard to add if we go forward with a `LineColumn` type.

Rather than using a more complicated struct to represent a location, we thought it'd be better to just have separate methods for `start_line()`, `start_column()`, `end_line()` etc.

Later, in a short discussion between @dtolnay and me, we concluded that the majority of use cases are about the start location, after which I suggested just having `.line()` and `.column()` (and maybe `.byte_offset()` in the future), just for the start of the span. In that case, you can still use `.after()` (the function returning the empty span after a span) to access the end line and column.

So, proposal:

- Remove the `LineColumn` type.
- Remove `span.start()` and `span.end()`.
- Maybe rename `span.before()` and `span.after()` to `span.start()` and `span.end()` now that those names are available again.
- Add `span.line()` and `span.column()`, with a note in the documentation to point out the use of `span.after().line()` (or `span.end().line()`) for the location of the end of the span.

**Comment by jhpratt** [comment]
That sounds great to me as well. Full for support of byte offset, as it would be necessary to construct subspans of existing tokens.

I can put together a PR tonight for what's stated (aka no byte offset), more than likely, if that's desirable?

**Comment by Zagitta** [comment]
Any chance `.byte_offset()` could be included initially? 

My "mad scientist" use case is to use track a struct back through time with the help of git and generate SQL migrations based on the evolution.
Since the libgit2 API uses `content_offset()` which is bytes I really need to know the byte offset to do the proper tracking.

**Comment by jhpratt** [comment]
If the information is already tracked (I think it is, but not positive), I can include it. It'll be up to the reviewer, but as it's nightly I think it would be okay.

**Comment by m-ou-se** [comment]
> Could you propose the resulting Public API here, like in the initial comment, to see how it all looks?

The proposal implemented in https://github.com/rust-lang/rust/pull/111571 is:

```rust
impl Span {
    /// The line number where the span starts.
    ///
    /// (Use .end().line() for the end line.)
    pub fn line(&self) -> usize;

    /// The column number where the spans starts.
    ///
    /// (Use .end().column() for the end column.)
    pub fn column(&self) -> usize;

    /// A zero width span at the start of this span.
    pub fn start(&self) -> Span;

    /// A zero width span at the end of this span.
    pub fn end(&self) -> Span;
}
```

(The `.start()` and `.end()` spans are also useful for diagnostics, e.g. to point at a missing keyword or character before/after something.)

How to expose the file name is still an open question. Currently, that is exposed as:

```rust
impl Span {
    pub fn source_file(&self) -> SourceFile;
}
```

Where `SourceFile` has a `.path()` and a `.is_real()`. Perhaps we want something different.

For the byte offset, we currently have:

```rust
impl Span {
    pub fn byte_range(&self) -> Range<usize>;
}
```

But since we don't have line_range and column_range either, it might make more sense to have byte_offset and byte_length that both just return a usize. But I don't have a strong opinion on this.

I think the `eq`, `join`, `parent` and `source` functions require a bit more work and/or discussion before those can possibly be stabilized.

**Comment by andrewdavidmackenzie** [comment]
I am interested in the `source_file()` method. I am not sure what case the `is_real() = false` covers (that can't be checked with a `Path`), but presumably there is one... I'll take whatever I can get TBH, as want to get off nightly and this is my blocker.

**Comment by m-ou-se** [comment]
I believe `source_file()` currently exposes the 'mapped' file name, when that feature is used. That is the file name that should be displayed in e.g. a panic message, but that's not the actual file name on disk that some proc macros need to access files at relative paths. (E.g. `c_include!("thing.h")`.)

The cases where `.is_real()` currently gives `false` have nothing to do with mapped file names, but are the cases where one does e.g. `TokenStream::from_str("..")`.

@dtolnay and I briefly discussed this moments ago, and we think that `SourceFile` should have both a method for the mapped path and for the actual on-disk path. The former should be used in panic/debug messages, and so on. The latter should be used when passing the file name to some external tool or when opening relative paths.

For 'fake' files, perhaps the "actual file path" method could just write the contents to a temporary file and return that path, to avoid having to use an `Option`. (The situation where a from_str-created TokenStream invokes a proc macro that uses the actual path seems very very rare anyway.)

(Suggestions for method names welcome. ^^)

**Comment by andrewdavidmackenzie** [comment]
The `is_real()`case you explain makes sense.

Maybe not a great name?
Could be covered by `source_file()` returning `Option<Path>` or similar? 
Will return `None`, when no file was used?

To clarify, my use case requires the path to the file on disk as mentioned.

**Comment by est31** [comment]
Re the proposal [above](https://github.com/rust-lang/rust/issues/54725#issuecomment-1546918161) by @m-ou-se:

> Later, in a short discussion between @dtolnay and me, we concluded that the majority of use cases are about the start location

Which use cases would those be? I think for diagnostics ideally one should use APIs provided by the compiler, and even if you make your own diagnostics printing library, ypu might be interested in the end line/column so that you can adjust the source printing.

I think a `LineColumn` like API is great, even though adding the relative byte offset would require changing the name to e.g. `Position` instead.

Also, before stabilization one should make the lines and maybe also columns relative: https://github.com/rust-lang/rust/issues/54725#issuecomment-1143966021 .

**Comment by sam0x17** [comment]
Here are some use cases that I think could inform some decisions here:

Right now in my [docify](https://crates.io/crates/docify), this is a big pain to deal with. What docify does is it lets you embed items (marked with an attribute macro) from throughout your project as examples in rust-docs that get copied in at compile-time. With span line and column info, this would have been extremely easy to pull off, because I would just have to syn parse the file, find my attribute, grab the span for the item it is attached to, and use those line and column numbers to grab the actual source for that item from the original input string. Instead I have to do this super hacky stuff that I wish I didn't have to do: https://github.com/sam0x17/docify/blob/main/macros/src/lib.rs#L456-L535

Here's another one: Consider people writing templating systems in rust where you can have rust syntax embedded in things like markdown or html files surrounded by `<?` tags, as is custom for these sorts of things. If it were possible to get line and column info from the rust excerpts, it would be possible to translate this info to be relative to the enclosing larger file. Side note there, would be nice to be able to _create_ a span manually for an arbitrary file for that use case, though that is a whole other issue.

I just want to illustrate that there are indeed some very eager potential users of this stuff

**Comment by Jules-Bertholet** [comment]
@m-ou-se's [`inline-python`](https://github.com/fusion-engineering/inline-python) library is another example.

**Comment by m-ou-se** [comment]
One example I can think of is a proc macro that calls some external tool or library and provides it line number information. E.g. inserting [`#line` directives](https://gcc.gnu.org/onlinedocs/cpp/Line-Control.html) for a C preprocessor or assembler.

If the proc macro itself wants to produce diagnostics, it should ideally use the proc macro diagnostics api, but I imagine there are many cases where an external tool/library will be the one providing diagnostics (perhaps even at a later point in time), in which case it should be informed of the original source location.

**Comment by m-ou-se** [comment]
Another example is a library that provide debug/assert/panic macros, which might want to report the line and column as part of the message. Today they can already expand to something with line!() and column!(), but the proc macro might need to be access the values directly for certain optimizations/tricks (e.g. encoding it and/or writing the messages to a separate file or linker section or something).

**Comment by est31** [comment]
Those are valid use cases, although at least docify could use the stable `Span::source_text`, I think. inline-python might not be able to if it passes on the proper span to python (I haven't checked). Thanks!

And yeah, most of those use cases only need the start span.

**Comment by sam0x17** [comment]
> Those are valid use cases, although at least docify could use the stable `Span::source_text`, I think. inline-python might not be able to if it passes on the proper span to python (I haven't checked). Thanks!
> 
> And yeah, most of those use cases only need the start span.

`source_text` didn't work because it mangles spacing and things. For a while I was using that and then running the result through `prettyplease`, but the problem with that is `source_text` as well as `.to_token_stream().to_string()` both strip non-doc comments, which is a no-no for this use case, so had to roll my own system.

I did contemplate doing some sort of transformation where I convert all line and multiline comments to a puesdo attribute syntax like we already do for doc comments, so they get retained as AST nodes that I can then convert back to regular comments, but this doesn't work because comments can appear in positions that attributes can't appear in (i.e. not attached to any particular item).

**Comment by petrochenkov** [comment]
@sam0x17 
>but the problem with that is `source_text` as well as `.to_token_stream().to_string()` both strip non-doc comments

Are you sure that `source_text` does that?
It should not.

It literally looks into source code and selects its part corresponding to the span positions with all whitespace and comments, and without any pre- or post-processing.

(`.to_token_stream().to_string()` does strip whitespace and comments, because it works with tokens and not with text.)

**Comment by sam0x17** [comment]
> @sam0x17
> 
> > but the problem with that is `source_text` as well as `.to_token_stream().to_string()` both strip non-doc comments
> 
> Are you sure that `source_text` does that? It should not.
> 
> It literally looks into source code and selects its part corresponding to the span positions with all whitespace and comments, and without any pre- or post-processing.
> 
> (`.to_token_stream().to_string()` does strip whitespace and comments, because it works with tokens and not with text.)

ah, apparently this was my issue -- no matter what `Item` I pass I get this error:

```
`Result::unwrap()` on an `Err` value: Error("expected `;`")
```

when calling:

```rust
item.span().source_text().unwrap();
```

**Comment by de-vri-es** [comment]
Hmm, currently `source_text()` is still underpowered on stable, because you can not get a `Span` covering an entire `TokenStream`.

There is no `TokenStream::span()` and `Span::join()` is still unstable. So getting the source text of more than a single `TokenTree` isn't possible right now. Or am I missing something?

**Comment by sam0x17** [comment]
My problem was actually this @de-vri-es: https://github.com/dtolnay/proc-macro2/issues/385

**Comment by warpfork** [comment]
Hello friends.  I'd love to have access to `source_file()` as well (even with the interesting caveats I see above).

Since I see some discussion about usecases, maybe I can briefly add mine: I'm attempting to write a test-generating macro which creates tests per some files it finds on the filesystem.  In doing this, I'd like to search the filesystem relative to where the user invoke the macro.

It seems like `proc_macro::Span::call_site().source_file().path()` is about the only option I can find for figuring where I am, at high enough resolution in order to do this.  (I've seen some other options floating around which involve environment variables, but these seem to generally not do what I need; `CARGO_MANIFEST_DIR` gets a path to the crate, but no further.  Also, env vars; blegh.)

I'd be perfectly happy to see functions to get both "real" on-disk paths and any "mapped" paths, as proposed in comment https://github.com/rust-lang/rust/issues/54725#issuecomment-1547921802 .  That said, I'd also probably advocate for finalizing and shipping the current behaviors, because they seem to have been around a while and been useful enough as-is; and because those other methods seem like they could be added in the future without much fuss.

Cheers!  Thank you all for your considerations.

**Comment by sam0x17** [comment]
Yes being able to discover the current file proc macros are _expanding in_ at compile-time would be a huge win for the macro ecosystem. For one it would then be possible to implement `include_str!` in pure proc macros since we'd be able to do file-relative paths to the current expansion

**Comment by Nemo157** [comment]
These recent comments about doing file-relative includes make me worried that exposing `source_file` before https://github.com/rust-lang/rfcs/pull/3200 will lead to more subtly broken proc-macros (though, I suppose given the many using manifest-relative paths already that ship may have sailed).

**Comment by sam0x17** [comment]
> These recent comments about doing file-relative includes make me worried that exposing `source_file` before [rust-lang/rfcs#3200](https://github.com/rust-lang/rfcs/pull/3200) will lead to more subtly broken proc-macros (though, I suppose given the many using manifest-relative paths already that ship may have sailed).

Yeah this is still the lesser evil for sure, right now my crate for example has to fall back to using _workspace_-relative paths like this in lieu of knowing the current file: https://github.com/sam0x17/docify/blob/main/macros/src/lib.rs#L33-L50

**Comment by frehberg** [comment]
The features "Span::source_file()" and "Span::start()" would be really helpful to improve rust-doc macro "aquamarine"

**Comment by sam0x17** [comment]
IMO the `source_file()` API could be as simple as `fn path(&self) -> Option<Path>`

It's dead simple and gives us the ecosystem win everyone is looking for, and is probably the least contentious version of this

**Comment by andrewdavidmackenzie** [comment]
+100 to @sam0x17 last comment. 

No need to agree on and implement _all_ the stable API initially, if we can agree on that one method - which would have a lot of bang for the buck.

**Comment by ogoffart** [comment]
This is the use case from [Slint](https://github.com/slint-ui/slint) GUI Crate with embedded DSL in a macro:

 - We would like to access the `source_file()` in order to load resources (images, or imported .slint files) relative to the current file.
   Currently we load the files relative to the CARGO_MANIFEST_PATH, but this is not great because it's confusing compared to other stuff, and it makes the integration with LSP and other tools harder.
   The suggested `path()` would work perfectly fine.
   
 - We would need the byte offset (or line column) in order to know whether two token touch each-other. 
   For example, I want to know if `foo-bar` was written like that or like `foo - bar` because our DSL (inspired from CSS) use camel case.
   Currently, we [parse the Debug implementation of Span](https://github.com/slint-ui/slint/blob/5b1f76496ac76dc1fe13959756913c8e84685c41/api/rs/macros/lib.rs#L26). But that's an ugly hack that is not
   reliable. So something like byte_offset of line/column is needed.  Alternative would be other way to figure that out (such as Span::join() combined with source_text).

**Comment by m-ou-se** [comment]
> IMO the `source_file()` API could be as simple as `fn path(&self) -> Option<Path>`
> 
> It's dead simple and gives us the ecosystem win everyone is looking for, and is probably the least contentious version of this

Would that give the mapped or unmapped path? See https://github.com/rust-lang/rust/issues/54725#issuecomment-1547921802

**Comment by sam0x17** [comment]
The two methods proposal seems reasonable 💯 

I think we should decide that one of them is `source_path` and one of them has a longer name. My assumption would be `source_path` would be a real file path that you can read, and `mapped_source_path` would be the other one.

I don't care at all about the naming though let's just stabilize something that gives us all access to these goodies

**Comment by Caellian** [comment]
## Source path

> worried that exposing `source_file` before [rust-lang/rfcs#3200](https://github.com/rust-lang/rfcs/pull/3200) will lead to more subtly broken proc-macros

I don't see relative imports breaking, but even if they do I don't think postponing stabilization of a 99% agreed upon feature over an RFC that hasn't been accepted yet over fear that people might use it wrong is warranted.

> * #[deprecated] path(&self) -> PathBuf { self.virtual_path() }

Why add a deprecated method? If you start typing "path", intellisense will suggest all methods containing "path" in their name. Bare `path` function is ambiguous given that it can have two meanings.

> `mapped_source_path` would be the other one.

I don't think that's clear enough. Anyway, to push things along I've created a [**POLL**](http://polljunkie.com/poll/xxynen/path-names) where people can pick multiple names they thing convey the meaning of the description above them and I hope we'll have an agreement on naming by the end of the year, it's been 2 years already.

If [`is_real`](https://doc.rust-lang.org/nightly/proc_macro/struct.SourceFile.html#method.is_real) is false I'd expect project manifest relative path returning `None`.

## Other

**eq**: Two spans are equal if their byte ranges are equal and their absolute system paths are equal. Tokens shouldn't use spans for equality though as that would provide better composability.

**join**:  I don't see any signature other than `fn join(&self, other: &Self) -> Option<Self>` making sense - if they don't overlap it's None, otherwise minmax of both.

## Use cases

I mostly plan on doing evil things with these:
- Partial/fuzzy type inference and transformation.
- Generating `Deref`-ed `repr(C)` struct for composed `repr(C)` structs so that `Deref` allows access to all fields nested fields.

**Comment by Nemo157** [comment]
> I don't see relative manual imports breaking

They are already broken, if you change the file they read without changing any source code cargo doesn't notice that and won't rebuild the crate. They are also relying on a limitation of the current proc-macro implementation in rustc: it does not cache the output as part of the incremental cache. Once caching is implemented they should not be rerun even when the source changes if that doesn't affect their input.

**Comment by menasheofd** [comment]
Hello! I'm interested in writing a proc macro that utilizes the feature of obtaining the caller's file location. Is there any new update on this issue?

**Comment by sam0x17** [comment]
> Hello! I'm interested in writing a proc macro that utilizes the feature of obtaining the caller's file location. Is there any new update on this issue?

the capabilities remain the same in stable. There are some very cursed patterns you can do if there is a unique string you know will only appear in one file in the crate, like my very cursed `macro-gpt` does here: https://github.com/sam0x17/macro-gpt/blob/main/src/lib.rs#L239-L272

In practice this works great, but it is intended more as an IDE-extension-living-in-a-proc-macro than a real proc macro you would keep in your source code.

**Comment by dbsxdbsx** [comment]
I come here when I need to write a procedural macro which could get the file path the macro is called, chatGpt give me the code like this:
```rust
#[proc_macro]
pub fn example_macro(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
    // Get the `Span` where the macro was called
    let span = proc_macro::Span::call_site();

    // Use the `Span` to get the `SourceFile`
    let source_file = span.source_file();

    // Get the path of the `SourceFile`
    let path: PathBuf = source_file.path();

    // Now you can use the file path for further processing
    // ...

    // Return the original input for this example
    input
}
```
Then fn `source_file` pointing me here, am I at the right place?

**Comment by wyatt-herkamp** [comment]
> I come here when I need to write a procedural macro which could get the file path the macro is called, chatGpt give me the code like this:
> 
> ```rust
> #[proc_macro]
> pub fn example_macro(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
>     // Get the `Span` where the macro was called
>     let span = proc_macro::Span::call_site();
> 
>     // Use the `Span` to get the `SourceFile`
>     let source_file = span.source_file();
> 
>     // Get the path of the `SourceFile`
>     let path: PathBuf = source_file.path();
> 
>     // Now you can use the file path for further processing
>     // ...
> 
>     // Return the original input for this example
>     input
> }
> ```
> 
> Then fn `source_file` pointing me here. Am I at the right place? If so, due to the API is unstable, is there a workaround?

`source_file` is unstable. Yes, a few workarounds exist. Such as what @sam0x17 said [here](https://github.com/sam0x17/macro-gpt/blob/main/src/lib.rs#L239-L272)

However, if you need something a little less sketchy I would just have your macro take an input of the source file location in the attributes. It would be less likely to break. 

You could also look at 
https://docs.rs/proc-macro2/latest/proc_macro2/#unstable-features

**Comment by sam0x17** [comment]
> However, if you need something a little less sketchy I would just have your macro take an input of the source file location in the attributes. It would be less likely to break.

Although in a multi-crate workspace this tends to break as well when you deploy to crates.io if any of the files are up one or more directories from the invocation, as I frequently encounter in my docify crate

**Comment by dbsxdbsx** [comment]
@wyatt-herkamp, @sam0x17 , thanks!  This workaround is feasible in my case, which wrapping a trait in a functional macro, and I modified it a little into this:
```rust
struct TraitVisitor {
    trait_name: String,
    found: Option<Macro>,
}

impl<'ast> Visit<'ast> for TraitVisitor {
    fn visit_macro(&mut self, mac: &'ast Macro) {
        if self.found.is_some() {
            return;
        }
        let last_seg = mac.path.segments.last().unwrap();
        if last_seg.ident != "trait_variable" {
            return;
        }

        // Convert the macro body tokens into a vector of Ident
        let idents: Vec<Ident> = mac.tokens.clone().into_iter().filter_map(|tt| match tt {
            proc_macro2::TokenTree::Ident(ident) => Some(ident),
            _ => None,
        }).collect();

        // Check for the presence of 'trait' keyword followed by the desired trait name
        for i in 0..idents.len() - 1 {
            if idents[i] == "trait" && idents[i + 1] == self.trait_name {
                println!("found trait: {:?}", self.trait_name);
                self.found = Some(mac.clone());
                break;
            }
        }
    }
}
```

**Comment by TheCataliasTNT2k** [comment]
Would be great to get the source file...

**Comment by Decodetalkers** [comment]
Emm. if without this feature, we cannot write something like include_str!() right?

**Comment by kanashimia** [comment]
> Emm. if without this feature, we cannot write something like include_str!() right?

No, this issue is unrelated. See two proposals mentioned here for doing that: https://github.com/rust-lang/rust/issues/54725#issuecomment-1345543702

**Comment by kanashimia** [comment]
# Embedding languages in a procedural macro

Here is my rant, it is been on my mind for a while:
I think a lot of _people here are misguided_, and _try to use macros in stupid ways_.
They try to embed languages that don't look like rust at all, and then complain.
For example some authors here mentioned previously the likes of slint, inline python, peg, whatever jsx crates.
YOU ARE DOING IT WRONG!!!!11!11
Proc macros take in RUST token trees, NOT PYTHON token trees and NOT SLINT token trees.
Rust macros operate on the RUST SYNTAX, stop complaining about the issues that come with that.
What YOU SHOULD DO instead is to make macros that take STRINGS as input.
Strings are the ONLY Rust constructs that can represent other languages in full. 
Take a look at what `asm!` "macro" does, it takes in strings, that is like the most standard example.
It may be ok to have some "non-standard" pieces of syntax for logging and whatnot, but don't do your whole macro like that.
Using macros for embedding languages not only breaks code highlighting completely, but also fundamentally contradicts with how proc macros work.
In comparison quasiquotes in Template Haskell actually take in string as an input instead of the token tree, imagine! Hence there are no problems with embedding something stupid like C++ in Haskell.

The real problem: 
When you embed a language as a string literal there is no stable way to make a meaningful span of the error, you can only make an error that spans the whole string.
`subspan` method on the literal should be stabilised for that, there is no need to try to hammer your nails with the `source_file` or `byte_range`, just stop. Using `subspan` is also much MUCH easier than trying to map between your DSL error positions and rust tokens and their positions. In the https://github.com/rust-lang/rust/issues/54725#issuecomment-1345543702 I show how it looks like. Unrelated but if Rust had a convention for highlighting stings it would even be possible to have an actual, meaningful, syntax highlighting, see https://github.com/rust-lang/rfcs/pull/3450#issuecomment-1806315495 where I was able to get a syntax highlighting for arbitrary strings by using comments.

TL;DR:
People should embed languages as string literals instead of arbitrary rust token trees in their proc macros.
`subspan` method on a literals needs to be stabilised first for better error reporting.

Proposal:
Maybe it would be meaningful to just make a minimal version of this proposal, called `proc_macro_subspan` and stabilise that, it would only contain `subspan` method, that is all that is needed for this use case.  

The only thing I'm not sure is why `subspan` is a method on the `Literal` and not on the `Span` instead. Are there any problems with the hygine or something? Having it on the `Span` would remove the need for `start()` and `end()` methods.

**Comment by workingjubilee** [comment]
> I have a concern about exposing `LineColum` information. It looks like it could play badly with incremental compilation, especially in the IDE context.
> 
> My understanding is that, if one adds a blank line to the start of the file, the line_column information of the input spans to proc macro changes. That means that IDE would have to re-expand procedural macros even after insignificant white space changes.

I believe that rustc now uses a span model that _is_ compatible with this ask, but currently doesn't use it during macro expansion, ironically.

**Comment by ethanuppal** [comment]
I don't believe it's possible to get the absolute path of a source file with `fn source_path()`? It's not possible to pass `env!(CARGO_MANIFEST_DIR)` through a procedural macro; returning a `PathBuf` relative to the manifest isn't as helpful in that regard.

**Comment by ogoffart** [comment]
I’d like to see progress on this issue. How can we help move it forward?  

### Source File  

My main use case is to locate file references in macros, similar to how `include_str!("file.txt")` resolves "file.txt" relative to its span. For Slint, this would help locate referenced images or other `.slint` files. Right now, resolving paths relative to the manifest is unintuitive.  
Another example of a use case is [here](https://github.com/alekseysidorov/include-utils?tab=readme-ov-file#notes)

It seems the main blocker for stabilizing `path()` is the need to rename it to `local_path()` to distinguish it from the mapped path. Should I propose an API Change Proposal (ACP) in the `libs-team` repo for this, or a directly make a PR? 
(`local_path` is how it is [currently named in rustc](https://github.com/rust-lang/rust/blob/98572840b6da350e5615a68586a766a0f9f1470a/compiler/rustc_span/src/lib.rs#L189))

### Line/Column Information  

Many use cases have been mentioned already. The main one i have in Slint is to know if an identifier token and a punctuation token are separated by space or not. Other usecases are using it with tools like [`xtr`](https://crates.io/crates/xtr) and [`cpp_build`](https://crates.io/crates/cpp_build) that parse code with `syn` and `proc_macro2` outside of macros, such as in CLI tools or build scripts. Line info would help with better diagnostics. Because even if not using the proc_macro API, the proc_macro2 crate only expose what is currently stable in proc_macro.

Regarding concerns about incremental compilation, I think this is not a real issue:  
- The Span API can track whether location methods are called, so the compiler can handle location dependencies only if needed.  
- Currently, macros already require re-runs, because they depends on extra files, environment variables, timestamps, network access, etc. So span location tracking would be one of the simpler issues to solve.  

Given this, I don’t see why line/column APIs couldn’t move toward stabilization. Are there other blockers preventing this?

**Comment by m-ou-se** [comment]
An open question is what to do with `SourceFile::is_real()`. It seems to always return `true`. Even in cases like `TokenStream::from_str`, the spans are just set to the call site with a real path.

Do we need `is_real`? Will it be used in the future? In which cases? Do we want to use an Option or some other interface for it instead?

**Comment by m-ou-se** [comment]
I believe the best thing we can do is to remove the `SourceFile` type entirely and instead have:

```rust
impl Span {
    pub fn file(&self) -> String; // Mapped file name, for display purposes.
    pub fn local_file(&self) -> Option<PathBuf>; // Real file name as it exists on disk.
}
```

PR: https://github.com/rust-lang/rust/pull/139671

**Comment by m-ou-se** [comment]
PR for stabilizing start+end+line+column+file+local_file, to be discussed by the libs-api team: https://github.com/rust-lang/rust/pull/139865

**Comment by m-ou-se** [comment]
After discussion in the libs-api meeting, I've split out the `file` and `local_file` from the stabilization PR, so we can first stablize start+end+line+column. Then we have some time to discuss local_file (and its name) and byte_range (or byte_offset).

**Comment by m-ou-se** [comment]
Opened https://github.com/rust-lang/rust/issues/139903 for the open question on the name of `local_file`. Feel free to leave a comment there (or react with 👍 if there's already comment you agree with).

**Comment by VictorArcium** [comment]
This should be updated now that `proc_macro_span_file` has been stabilized (in 1.88).

**Comment by andrewdavidmackenzie** [comment]
I hadn't noticed that had been stabalized - so thanks.
I was able to move my project (finally!) to stable.

**Comment by Aversefun** [comment]
What's the status of the `join` method? In other words, what still needs to be discussed about it?
